File: mobile/ios/Backplane/LogStore.swift
import Foundation

// The event frames a hub sent, kept per hub so the next launch shows the
// log at once and asks the hub only for what is new (since/origin), as
// the web client does with localStorage. One base64 frame per line: a full
// log ("reset") starts the file over, more changes ("append") add a line.
struct LogStore {
    static let max = 48 << 20

    private var dir: URL {
        let d = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0].appendingPathComponent("logs")
        try? FileManager.default.createDirectory(at: d, withIntermediateDirectories: true)
        return d
    }

    private func file(_ key: String) -> URL {
        let safe = String(key.map { $0.isLetter || $0.isNumber || $0 == "." || $0 == "-" || $0 == "_" ? $0 : "_" })
        return dir.appendingPathComponent(safe + ".b64")
    }

    func load(_ key: String) -> [String] {
        let f = file(key)
        guard let a = try? FileManager.default.attributesOfItem(atPath: f.path), let n = a[.size] as? Int else { return [] }
        // too big to replay quickly: start from the hub again
        if n > Self.max { try? FileManager.default.removeItem(at: f); return [] }
        guard let text = try? String(contentsOf: f, encoding: .utf8) else { return [] }
        return text.split(separator: "\n").map(String.init).filter { !$0.isEmpty }
    }

    func keep(_ key: String, _ how: String, _ frame: String) {
        let f = file(key)
        let line = Data((frame + "\n").utf8)
        switch how {
        case "reset":
            try? line.write(to: f, options: .atomic)
        case "append":
            if let h = try? FileHandle(forWritingTo: f) {
                h.seekToEndOfFile()
                h.write(line)
                try? h.close()
            }
        default:
            break
        }
    }

    func forget(_ key: String) { try? FileManager.default.removeItem(at: file(key)) }
}

File: mobile/android/app/src/main/java/dev/backplane/mobile/LogStore.kt
package dev.backplane.mobile

import java.io.File

// The event frames a hub sent, kept per hub so the next launch shows the
// log at once and asks the hub only for what is new (since/origin), as
// the web client does with localStorage. One base64 frame per line: a full
// log ("reset") starts the file over, more changes ("append") add a line.
class LogStore(root: File) {
    private val dir = File(root, "logs").apply { mkdirs() }

    private fun file(key: String) = File(dir, key.replace(Regex("[^A-Za-z0-9._-]"), "_") + ".b64")

    fun load(key: String): List<String> {
        val f = file(key)
        if (!f.exists()) return emptyList()
        // too big to replay quickly: start from the hub again
        if (f.length() > MAX) { f.delete(); return emptyList() }
        return f.readLines().filter { it.isNotBlank() }
    }

    fun keep(key: String, how: String, frame: String) {
        val f = file(key)
        when (how) {
            "reset" -> f.writeText(frame + "\n")
            "append" -> if (f.exists()) f.appendText(frame + "\n")
        }
    }

    fun forget(key: String) { file(key).delete() }

    companion object { const val MAX = 48L shl 20 }
}
